modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -129,6 +130,245 @@ func (suite *KeeperTestSuite) TestChanUpgradeInit() {
+				path.EndpointA.GetChannel().UpgradeSequence,
+				proofCounterpartyChannel,
+				proofCounterpartyUpgrade,
+				proofHeight)
+
+			if tc.expPass {
+				suite.Require().NoError(err)
+				// suite.Require().Equal(expSequence, path.EndpointB.GetChannel().UpgradeSequence)
+				// suite.Require().Equal(mock.Version, path.EndpointB.GetChannel().Version)
+				// suite.Require().Equal(path.EndpointB.GetChannel().State, types.TRYUPGRADE)
+			} else {
+				suite.Require().Error(err)
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestChanUpgradeTry_CrossingHellos() {
+	var (
+		path *ibctesting.Path
+		// expSequence                 uint64
+		upgrade                     types.Upgrade
+		counterpartyUpgrade         types.Upgrade
+		counterpartyUpgradeSequence uint64
+	)
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expPass  bool
+	}{
+		{
+			"success",
+			func() {},
+			true,
+		},
+		// {
+		// 	"success: counterparty sequence > channel.UpgradeSequence",
+		// 	func() {
+		// 		counterpartyUpgradeSequence = 5
+		// 		expSequence = 5
+		// 	},
+		// 	true,
+		// },
+		// {
+		// 	"fail: upgrade fields have changed",
+		// 	func() {
+		// 		counterpartyUpgrade.Fields.Ordering = types.ORDERED
+		// 		counterpartyUpgrade.Fields.Version = fmt.Sprintf("%s-v3", mock.Version)
+		// 	},
+		// 	false,
+		// },
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			// chainA UpgradeInit
+			path.EndpointA.ChannelConfig.ProposedUpgrade.Fields.Version = fmt.Sprintf("%s-v2", mock.Version)
+			err := path.EndpointA.ChanUpgradeInit()
+			suite.Require().NoError(err)
+
+			// commit a block to update chain A for correct proof querying
+			path.EndpointA.Chain.Coordinator.CommitBlock(path.EndpointA.Chain)
+			// update chainB's client of chain A to account for ChanUpgradeInit
+			suite.Require().NoError(path.EndpointB.UpdateClient())
+
+			// we also UpgradeInit to simulate crossing hellos situation
+			path.EndpointB.ChannelConfig.ProposedUpgrade.Fields.Version = fmt.Sprintf("%s-v2", mock.Version)
+			err = path.EndpointB.ChanUpgradeInit()
+			suite.Require().NoError(err)
+
+			// update chainA's client of chain B to account for ChanUpgradeInit
+			suite.Require().NoError(path.EndpointA.UpdateClient())
+
+			counterpartyUpgrade = path.EndpointA.GetProposedUpgrade()
+			proposedConnectionHops := []string{path.EndpointB.ConnectionID}
+			upgrade = types.NewUpgrade(
+				types.NewUpgradeFields(
+					types.UNORDERED, proposedConnectionHops, fmt.Sprintf("%s-v2", mock.Version),
+				),
+				types.NewTimeout(path.EndpointA.Chain.GetTimeoutHeight(), 0),
+				0,
+			)
+
+			tc.malleate()
+
+			// we need to update the clients again because malleation has changed the channel state
+			suite.Require().NoError(path.EndpointA.UpdateClient())
+			suite.Require().NoError(path.EndpointB.UpdateClient())
```

Comment:
same reasoning as above, the setup code only needs to update the chainB client once